core/src/main/java/jenkins/security/ClassFilterImpl.java

```diff
@@ -0,0 +1,257 @@
+        } catch (IOException x) {
+            throw new ExceptionInInitializerError(x);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    public boolean isBlacklisted(Class _c) {
+        for (CustomClassFilter f : ExtensionList.lookup(CustomClassFilter.class)) {
+            Boolean r = f.permits(_c);
+            if (r != null) {
+                LOGGER.log(Level.FINER, "{0} specifies a policy for {1}: {2}", new Object[] {f, _c.getName(), r});
+                return !r;
+            }
+        }
+        return cache.computeIfAbsent(_c, c -> {
+            if (ClassFilter.STANDARD.isBlacklisted(c)) { // currently never true: only the name overload is overridden
+                return true;
+            }
+            String name = c.getName();
+            if (Main.isUnitTest && name.contains("$$EnhancerByMockitoWithCGLIB$$")) {
+                mockOff();
+                return false;
+            }
+            if (c.isArray()) {
+                LOGGER.log(Level.FINE, "permitting {0} since it is an array", name);
+                return false;
+            }
+            if (Throwable.class.isAssignableFrom(c)) {
+                LOGGER.log(Level.FINE, "permitting {0} since it is a throwable", name);
+                return false;
+            }
+            if (Enum.class.isAssignableFrom(c)) { // Class.isEnum seems to be false for, e.g., java.util.concurrent.TimeUnit$6
+                LOGGER.log(Level.FINE, "permitting {0} since it is an enum", name);
+                return false;
+            }
+            CodeSource codeSource = c.getProtectionDomain().getCodeSource();
+            URL location = codeSource != null ? codeSource.getLocation() : null;
+            if (location != null) {
+                if (isLocationWhitelisted(location.toString())) {
+                    LOGGER.log(Level.FINE, "permitting {0} due to its location in {1}", new Object[] {name, location});
+                    return false;
+                }
+            } else {
+                ClassLoader loader = c.getClassLoader();
+                if (loader != null && loader.getClass().getName().equals("hudson.remoting.RemoteClassLoader")) {
+                    LOGGER.log(Level.FINE, "permitting {0} since it was loaded by a remote class loader", name);
+                    return false;
+                }
+            }
+            if (WHITELISTED_CLASSES.contains(name)) {
+                LOGGER.log(Level.FINE, "tolerating {0} by whitelist", name);
+                return false;
+            }
+            LOGGER.log(Level.WARNING, "{0} in {1} might be dangerous, so rejecting; see https://jenkins.io/redirect/class-filter/", new Object[] {name, location != null ? location : "JRE"});
+            return true;
+        });
+    }
+
+    private static final Pattern CLASSES_JAR = Pattern.compile("(file:/.+/)WEB-INF/lib/classes[.]jar");
```

Comment:
NIT: code conventions is static fields are declared before non static fields.

Reply:
I wanted to keep this close to its sole usage.